teeworlds7/networking.go

```diff
@@ -14,71 +17,121 @@ const (
 
-func readNetwork(ch chan<- []byte, conn net.Conn) {
 	buf := make([]byte, maxPacksize)
-
 	for {
-		len, err := bufio.NewReader(conn).Read(buf)
-		packet := make([]byte, len)
-		copy(packet[:], buf[:])
-		if err == nil {
-			ch <- packet
-		} else {
-			fmt.Printf("Some error %v\n", err)
-			break
+		n, err := conn.Read(buf)
+		if err != nil {
+			cancelCause(err)
+			return
+		}
+
+		if n == 0 {
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				continue
+			}
+		}
+
+		packet := make([]byte, n)
+		copy(packet, buf[:n])
+		select {
+		case ch <- packet:
+		case <-ctx.Done():
+			return
 		}
 	}
+}
 
-	conn.Close()
+func (client *Client) Connect(serverIp string, serverPort int) error {
+	return client.ConnectContext(context.Background(), serverIp, serverPort)
 }
 
-func (client *Client) Connect(serverIp string, serverPort int) {
-	ch := make(chan []byte, maxPacksize)
+func (client *Client) ConnectContext(ctx context.Context, serverIp string, serverPort int) (err error) {
+	ctx, cancelCause := context.WithCancelCause(ctx)
+	defer func() {
+		// can be called only once
+		// in case this cause is second, it will not override
+		// the initial cancelation cause
+		cancelCause(nil)
+
+		ctxErr := context.Cause(ctx)
+		if ctxErr != nil {
+			err = ctxErr
+			return
+		}
+
+		ctxErr = ctx.Err()
+		if ctxErr != nil && !errors.Is(ctxErr, context.Canceled) {
+			err = ctxErr
+			return
+		}
+	}()
 
-	conn, err := getConnection(serverIp, serverPort)
+	ch := make(chan []byte, maxPacksize)
+	var d net.Dialer
+	conn, err := d.DialContext(ctx, "udp", fmt.Sprintf("%s:%d", serverIp, serverPort))
 	if err != nil {
-		fmt.Printf("error connecting %v\n", err)
-		return
+		return fmt.Errorf("failed to connect to server: %s:%d: %w", serverIp, serverPort, err)
 	}
+	client.Conn = conn
+	defer func() {
+		closeErr := client.Conn.Close()
+		if closeErr != nil {
+			slog.Error("failed to close connection", "error", closeErr)
+		}
+	}()
 
-	client.Session = protocol7.Session{
-		ClientToken: [4]byte{0x01, 0x02, 0x03, 0x04},
-		ServerToken: [4]byte{0xff, 0xff, 0xff, 0xff},
-		Ack:         0,
-	}
+	client.Session = protocol7.NewSession()
 	client.Game.Players = make([]Player, network7.MaxClients)
-	client.Conn = conn
 
-	go readNetwork(ch, conn)
+	var wg sync.WaitGroup
+	wg.Add(1)
+	defer wg.Wait() // wait for the reader goroutine to finish execution, before leaving this function scope
 
-	client.SendPacket(client.Session.CtrlToken())
+	go readNetwork(ctx, cancelCause, &wg, ch, conn)
+
+	err = client.SendPacket(client.Session.CtrlToken())
+	if err != nil {
+		return fmt.Errorf("failed to send token: %w", err)
+	}
 
 	// TODO: do we really need a non blocking network read?
 	//       if not remove the channel, the sleep and the select statement
 	//       if yes also offer an OnTick callback to the user, and also do keepalives and resends
+
+	ticker := time.NewTicker(10 * time.Millisecond)
+	defer ticker.Stop()
+
 	for {
-		time.Sleep(10_000_000)
```

Comment:
Does removing this increase the CPU usage?

Reply:
this is replaced by the ticker.
The ticker has an overhead which allows it to be stopped and it provides a channel, which allows us to wait for the "sleep" in parallel to waiting for a signal.

I did not check cpu usage

Reply:
A slight overhead from a smarter ticker is fine. But if this causes fan spin or 100% cpu usage thats not acceptable.

Reply:
should not be the case.